File: solutions/ocaml/121-lowest-common-ancestor-of-a-binary-tree.ml
(*
  Problem 121: Lowest Common Ancestor of a Binary Tree (LeetCode 236)
  Difficulty: Med
  Language: OCaml
*)
type 'a btree = BLeaf | BNode of 'a btree * 'a * 'a btree

let rec lca root p q = match root with
  | BLeaf -> None
  | BNode (_, v, _) when v = p || v = q -> Some v
  | BNode (l, v, r) ->
    match lca l p q, lca r p q with
    | Some _, Some _ -> Some v
    | Some x, None | None, Some x -> Some x
    | _ -> None

File: solutions/haskell/121-lowest-common-ancestor-of-a-binary-tree.hs
{-
  Problem 121: Lowest Common Ancestor of a Binary Tree (LeetCode 236)
  Difficulty: Med
  Language: Haskell
-}
data BTree a = BLeaf | BNode (BTree a) a (BTree a) deriving Eq

lca :: Eq a => BTree a -> a -> a -> Maybe a
lca BLeaf _ _ = Nothing
lca (BNode l v r) p q
  | v == p || v == q = Just v
  | otherwise = case (lca l p q, lca r p q) of
      (Just _, Just _) -> Just v
      (Just x, Nothing) -> Just x
      (Nothing, Just x) -> Just x
      _ -> Nothing
